exsim/ltem.py

```diff
@@ -0,0 +1,332 @@
+    wavelength : numbers.Real, optional
+        Relativistic wavelength of the electrons in m.
+
+    Returns
+    -------
+    discretisedfield.Field
+        Intensity at specified defocus.
+
+    Examples
+    --------
+
+    1. Zero defocus.
+
+    >>> import discretisedfield as df
+    >>> import exsim
+    >>> import numpy as np
+    >>> mesh = df.Mesh(p1=(-5e-9, -4e-9, -1e-9), p2=(5e-9, 4e-9, 1e-9),
+    ...                    cell=(2e-9, 1e-9, 0.5e-9))
+    ...
+    >>> def value_fun(point):
+    ...     x, y, z = point
+    ...     if x > 0:
+    ...         return (0,-1, 0)
+    ...     else:
+    ...         return (0, 1, 0)
+    ...
+    >>> field = df.Field(mesh, dim=3, value=value_fun)
+    >>> phase, ft_phase = exsim.ltem.phase(field)
+    >>> df_img = exsim.ltem.defocus_image(phase, Cs=0, df_length=0,
+    ...                                   U=300e3)
+    >>> np.allclose(df_img.array, [1])
+    True
+
+    .. plot::
+        :context: close-figs
+
+        2. Visualising the phase using ``matplotlib``.
+
+        >>> import discretisedfield as df
+        >>> import exsim
+        >>> mesh = df.Mesh(p1=(-5e-9, -4e-9, -1e-9), p2=(5e-9, 4e-9, 1e-9),
+        ...                    cell=(2e-9, 1e-9, 0.5e-9))
+        ...
+        >>> def value_fun(point):
+        ...     x, y, z = point
+        ...     if x > 0:
+        ...         return (0, -1, 0)
+        ...     else:
+        ...         return (0, 1, 0)
+        ...
+        >>> field = df.Field(mesh, dim=3, value=value_fun)
+        >>> phase, ft_phase = exsim.ltem.phase(field)
+        >>> df_img = exsim.ltem.defocus_image(phase, Cs=8000, df_length=0.2e-3,
+        ...                                   U=300e3)
+        >>> df_img.mpl_scalar()
+
+    .. seealso::
+
+        :py:func:`~exsim.ltem.phase`
+        :py:func:`~exsim.ltem.relativistic_wavelength`
+
+    """
+    # Perhaps we can add more functionality to discretisedfield to simplify
+    # these operations. The following line would then be: df.exp(phase *
+    # 1j).fft2(). The missing functionalities: (1) multiplication with complex
+    # scalars, (2) df.exp(). For consistency, we could also add df.sim(),
+    # df.cos(), df.sqrt(), etc. to mimic numpy's behaviour.
+    ft_wavefn = df.Field(phase.mesh, dim=phase.dim,
+                         value=np.exp(phase.array * 1j)).fft2()
+    k = df.Field(ft_wavefn.mesh, dim=3, value=lambda x: x)
+    ksquare = (k.x**2 + k.y**2).array  # Can we do this without exposing array?
+
+    if wavelength is None:
+        if U is None:
+            msg = ('Either `wavelength` or acceleration voltage `U` needs'
+                   'to be specified.')
+            raise RuntimeError(msg)
+        wavelength = relativistic_wavelength(U)
+
+    cts = -df_length + 0.5 * wavelength**2 * Cs * ksquare  # This could probably work with ksquare being an array
+    exp = np.exp(np.pi * cts * 1j * ksquare * wavelength)  # Similarly, if we had df.exp, this could be simplified.
```

Comment:
Not exposing `array` from `ksquare` will only work if we have `df.exp`. For now `cts` and `ksquare` must be numpy arrays.